Source/Particles/ParticleCreation/FilterCopyTransform.H

```diff
@@ -0,0 +1,116 @@
+#ifndef FILTER_COPY_TRANSFORM_H_
+#define FILTER_COPY_TRANSFORM_H_
+
+#include <AMReX_GpuContainers.H>
+#include <AMReX_TypeTraits.H>
+
+/**
+ * \brief Apply a filter, copy, and transform operation to the particles
+ * in src, in that order, writing the result to dst, starting at dst_index.
+ * The dst tile will be extended so all the particles will fit, if needed.
+ *
+ * Note that the transform function operates on both the src and the dst,
+ * so both can be modified.
+ *
+ * \tparam DstTile the dst particle tile type
+ * \tparam SrcTile the src particle tile type
+ * \tparam Index the index type, e.g. unsigned int
+ * \tparam TransFunc the transform function type
+ * \tparam CopyFunc the copy function type
+ *
+ * \param dst the destination tile
+ * \param src the source tile
+ * \param mask pointer to the mask - 1 means copy, 0 means don't copy
+ * \param dst_index the location at which to starting writing the result to dst
+ * \param copy defines the function that will be applied for the "copy" step
+ * \param transform defines that transformation.
+ *
+ */
+template <typename DstTile, typename SrcTile, typename Index,
+          typename TransFunc, typename CopyFunc,
+          amrex::EnableIf_t<std::is_integral<Index>::value, int> foo = 0>
+Index filterCopyTransformParticles (DstTile& dst, SrcTile& src, Index* mask, Index dst_index,
```

Comment:
Which of these params can be promoted to `const` for const-correctness?
`mask` and  `dst_index`?
Keeping `src` a regular reference is good for flexibility, e.g. ionization changes the source.